src/common/src/array/list_array.rs

```diff
@@ -429,22 +449,19 @@ impl<'a> ListRef<'a> {
         &self,
         serializer: &mut memcomparable::Serializer<impl BufMut>,
     ) -> memcomparable::Result<()> {
+        let mut inner_serializer = memcomparable::Serializer::new(vec![]);
         match self {
-            ListRef::Indexed { arr, idx } => {
-                let offsets = arr.offsets[*idx]..arr.offsets[*idx + 1];
-                serializer.serialize_u32(offsets.len() as u32)?;
-                offsets
-                    .map(|o| arr.value.value_at(o))
-                    .try_for_each(|datum_ref| serialize_datum_ref_into(&datum_ref, serializer))?
-            }
-            ListRef::ValueRef { val } => {
-                serializer.serialize_u32(val.values.len() as u32)?;
-                val.values
-                    .iter()
-                    .try_for_each(|datum| serialize_datum_into(datum, serializer))?
-            }
+            ListRef::Indexed { arr, idx } => (arr.offsets[*idx]..arr.offsets[*idx + 1])
+                .map(|o| arr.value.value_at(o))
+                .try_for_each(|datum_ref| {
+                    serialize_datum_ref_into(&datum_ref, &mut inner_serializer)
+                })?,
+            ListRef::ValueRef { val } => val
+                .values
+                .iter()
+                .try_for_each(|datum| serialize_datum_into(datum, &mut inner_serializer))?,
```

Comment:
You can use the `values_ref` function to iterate the elements. Actually, we can do a slight refactoring to it and make it efficient for iterating:

```
fn values_ref(&self) -> Vec<DatumRef<'a>> {
  self.iter_values().collect()
}

fn iter_values() -> Iterator<DatumRef<'a>> {
  ...
}
``` 


```
    pub fn serialize(
        &self,
        serializer: &mut memcomparable::Serializer<impl BufMut>,
    ) -> memcomparable::Result<()> {
      self.iter_values_().try_for_each(|d| ...)
    }
```

Reply:
For different variants, the concrete type of the iterator is different, so we may need to introduce `auto_enum` here which needs to match the variant on every call of `next`. Not sure whether this refactoring is worthwhile.

Reply:
From tests we can see that currently we support nested array, e.g. `[[1, 2], [2, 3]]`. Is memcomparable really suitable for this case?